spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java

```diff
@@ -167,43 +206,156 @@ boolean isNonEntityReturnedType(Class returnedType) {
 			List<String> tags,
-			Object[] vals) {
+			List vals) {
 		Builder builder = GqlQuery.newGqlQueryBuilder(gql);
 		builder.setAllowLiteral(true);
-		if (tags.size() != vals.length) {
+		if (tags.size() != vals.size()) {
 			throw new DatastoreDataException("Annotated GQL Query Method "
 					+ this.queryMethod.getName() + " has " + tags.size()
-					+ " tags but a different number of parameter values: " + vals.length);
+					+ " initialTags but a different number of parameter values: " + vals.size());
 		}
 		for (int i = 0; i < tags.size(); i++) {
-			Object val = vals[i];
+			Object val = vals.get(i);
 			DatastoreNativeTypes.bindValueToGqlBuilder(builder, tags.get(i), val);
 		}
 		return builder.build();
 	}
+
+	private void resolveSpelTags(QueryTagValue queryTagValue) {
+		Expression[] expressions = getSpelExpressionParts();
+		StringBuilder sb = new StringBuilder();
+		Map<Object, String> valueToTag = new HashMap<>();
+		int tagNum = 0;
+		EvaluationContext evaluationContext = this.evaluationContextProvider
+				.getEvaluationContext(this.queryMethod.getParameters(),
+						queryTagValue.rawParams);
+		for (Expression expression : expressions) {
+			if (expression instanceof LiteralExpression) {
+				sb.append(expression.getValue(String.class));
+			}
+			else if (expression instanceof SpelExpression) {
+				Object value = expression.getValue(evaluationContext);
+				if (valueToTag.containsKey(value)) {
+					sb.append("@").append(valueToTag.get(value));
+				}
+				else {
+					String newTag;
+					do {
+						tagNum++;
+						newTag = "SpELtag" + tagNum;
+					}
+					while (queryTagValue.initialTags.contains(newTag));
+					valueToTag.put(value, newTag);
+					queryTagValue.params.add(value);
+					queryTagValue.tagsOrdered.add(newTag);
+					sb.append("@").append(newTag);
+				}
+			}
+			else {
+				throw new DatastoreDataException(
+						"Unexpected expression type. GQL queries are expected to be "
+								+ "concatenation of Literal and SpEL expressions.");
+			}
+		}
+		queryTagValue.finalGql = sb.toString();
+	}
+
+	private Expression[] getSpelExpressionParts() {
+		if (this.spelExpressionParts == null) {
+			this.spelExpressionParts = detectExpressions(getGqlResolvedEntityClassName());
+		}
+		return this.spelExpressionParts;
+	}
+
+	private Expression[] detectExpressions(String gql) {
+		Expression expression = this.expressionParser.parseExpression(gql,
+				ParserContext.TEMPLATE_EXPRESSION);
+		if (expression instanceof LiteralExpression) {
+			return new Expression[] { expression };
+		}
+		else if (expression instanceof CompositeStringExpression) {
+			return ((CompositeStringExpression) expression).getExpressions();
+		}
+		else {
+			throw new DatastoreDataException("Unexpected expression type. "
+					+ "Query can either contain no SpEL expressions or have only "
+					+ "literal SpEL expressions in the GQL.");
+		}
+	}
+
+	private String resolveEntityClassNames(String sql) {
```

Comment:
add comments or java doc

Reply:
We don't javadoc private methods, but I can add a comment.